File: p2p-example.go
package main

import (
	"bufio"
	"context"
	"fmt"
	"log"
	"strings"

	libp2p "github.com/libp2p/go-libp2p"
	peer "github.com/libp2p/go-libp2p-core/peer"
	protocol "github.com/libp2p/go-libp2p-core/protocol"
)

func main() {
	// Create a libp2p host with a random peer ID and listen on a random port
	host, err := libp2p.New(context.Background())
	if err != nil {
		log.Fatal(err)
	}
	defer host.Close()

	// Parse the peer ID of the other host from the command-line arguments
	otherPeerID, err := peer.IDB58Decode(strings.TrimSpace(string(os.Args[1])))
	if err != nil {
		log.Fatal(err)
	}

	// Print the host's peer ID and listen address
	fmt.Printf("Peer ID: %s\n", host.ID())
	fmt.Printf("Listen address: /ip4/%s/tcp/%d\n", host.Addrs()[0].String(), host.Network().ListenAddresses()[0].TCPAddr().Port)

	// Set up a stream between the two hosts
	stream, err := host.NewStream(context.Background(), otherPeerID, protocol.ID("example-protocol"))
	if err != nil {
		log.Fatal(err)
	}
	defer stream.Close()

	// Send a message from one host to the other
	if _, err := fmt.Fprintln(stream, "Hello, world!"); err != nil {
		log.Fatal(err)
	}

	// Read the response from the other host
	response, err := bufio.NewReader(stream).ReadString('\n')
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Response: %s", response)
}

File: p2p-example.rs
use libp2p::{
    identity,
    multiaddr::{Multiaddr, Protocol},
    peer::PeerId,
    tokio_rustls::rustls::{Certificate, PrivateKey},
    MultiaddrExt,
};
use std::{
    env,
    error::Error,
    io::{self, BufRead, BufReader, Write},
    net::{IpAddr, SocketAddr},
};
use tokio::{
    io::{AsyncReadExt, AsyncWriteExt},
    net::TcpStream,
    prelude::*,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Create a libp2p host with a random peer ID and listen on a random port
    let local_key = identity::Keypair::generate_ed25519();
    let local_peer_id = PeerId::from(local_key.public());
    let transport = libp2p::tcp::TcpConfig::new().nodelay(true);
    let mut host = libp2p::Host::new(transport, local_peer_id, local_key)?;
    let listen_addr = "/ip4/0.0.0.0/tcp/0".parse::<Multiaddr>()?;
    host.listen_on(listen_addr).await?;

    // Parse the peer ID of the other host from the command-line arguments
    let other_peer_id = env::args()
        .nth(1)
        .expect("Please provide the ID of the other host as an argument")
        .parse::<PeerId>()?;

    // Print the host's peer ID and listen address
    println!("Peer ID: {}", host.peer_id());
    for addr in host.addresses() {
        println!("Listen address: {}", addr);
    }

    // Set up a stream between the two hosts
    let protocol = "/example-protocol/1.0.0".parse::<Protocol>()?;
    let mut stream = host.new_stream(context::Context::background(), &other_peer_id, protocol).await?;

    // Send a message from one host to the other
    writeln!(stream, "Hello, world!")?;

    // Read the response from the other host
    let mut reader = BufReader::new(&mut stream);
    let mut response = String::new();
    reader.read_line(&mut response)?;
    println!("Response: {}", response);

    Ok(())
}
